File: examples/dof2.cpp
#include <qpOASES.hpp>
#include <cmath>
#include "ros/ros.h"
#include "sensor_msgs/JointState.h"


/** Example for obtaining the velocity for 1DOF using the QProblem class. */
int main(int argc, char **argv)
{
    ros::init(argc, argv, "controller_1dof");
    ros::NodeHandle n;
    ros::Publisher joint_pub = n.advertise<sensor_msgs::JointState>("joint_state", 10);
    ros::Rate loop_rate(10);
    ros::Duration(0.5).sleep();

    USING_NAMESPACE_QPOASES;

    /* Setup data of first QP. */
    /* Weights. */
    float w1 = 1e-4;
    float w2 = 1e-4;
    float w3 = 1;
    /* Joint limits. */
    float q0_min = -0.05;
    float q0_max = 0.05;
    float q1_min = -0.15;
    float q1_max = 0.15;
    /* Links */
    float l1 = 0.1;
    float l2 = 0.2;
    float l3 = 0.3;
    /* Initial joint values. */
    float q0 = -0.05;
    float q1 = -0.15;
    /* Joint target. */
    float q_des = 0.6;
    /* Slack limits. */
    float e_min = -10;
    float e_max = 10;
    /* Velocity limits. */
    float v0_min = -0.01;
    float v0_max = 0.01;
    float v1_min = -0.1;
    float v1_max = 0.1;
    /* Acceleration limits. */
    float a0_min = -0.005;
    float a0_max = 0.005;
    float a1_min = -0.01;
    float a1_max = 0.01;
    /* Others */
    float precision= 1e-4;
    float p = 4;
    float q_eef = l1 +l2 +l3 + q0 + q1;
    float error = p*(q_des-q_eef);
    float vel_init = 0;

    int nWSR = 100;

    real_t H[3*3] = { w1, 0.0, 0.0,  0.0, w2, 0.0,  0.0, 0.0, w3 };
    real_t A[3*3] = { 1.0, 1.0, 1.0,  1.0, 0.0, 0.0,  0.0, 1.0, 0.0 };
    real_t g[3] = { 0.0, 0.0, 0.0 };
    real_t lb[3] = { v0_min, v1_min, e_min };
    real_t ub[3] = { v0_max, v1_max, e_max };
    real_t lbA[3] = { error, (q0_min - q0), (q1_min - q0) };
    real_t ubA[3] = { error, (q0_max - q0), (q1_max - q0) };

    /* Setting up QProblem object. */
    SQProblem example( 3,3 );
    Options options;
    options.setToReliable();
    options.printLevel = qpOASES::PL_HIGH;
    example.setOptions(options);


    printf( "Init pos = %g,  goal = %g, error = %g, q0 =%g, q1 = %g\n",
            q_eef,q_des,error,q0,q1);

    int i = 0;
    float limit = std::abs(error);;
    float condition;
    bool first = true;
    returnValue return_value;
    real_t Opt[3];

    /* ROS Message definition */
    sensor_msgs::JointState vel_pos;
    vel_pos.header.frame_id = "q0";
    vel_pos.position.push_back(q_eef);
    vel_pos.velocity.push_back(vel_init);
    vel_pos.effort.push_back(error);
    vel_pos.effort.push_back(0);
    vel_pos.header.stamp  = ros::Time::now();
    loop_rate.sleep();

    return_value = example.init( H,g,A,lb,ub,lbA,ubA, nWSR );

    if(return_value != qpOASES::SUCCESSFUL_RETURN)
    {
              std::cout << "Init of QP-Problem returned without success! ERROR MESSAGE: " <<
                qpOASES::MessageHandling::getErrorCodeMessage(return_value) << std::endl;
              return -1;
    }


    while (ros::ok()) {
        do {

            if (first == true) {
                joint_pub.publish(vel_pos);
            }
            else {
                /* Solve QP. */
                ++i;
                nWSR = 100;
                lbA[0] = error;
                lbA[1] = q0_min - q0;
                lbA[2] = q1_min - q1;
                ubA[0] = error;
                ubA[1] = q0_max - q0;
                ubA[2] = q1_max - q1;

                return_value = example.hotstart( H,g,A,lb,ub,lbA,ubA, nWSR );

                if(return_value != qpOASES::SUCCESSFUL_RETURN)
                {
                          std::cout << "Hotstart of QP-Problem returned without success! ERROR MESSAGE: " <<
                            qpOASES::MessageHandling::getErrorCodeMessage(return_value) << std::endl;
                          return -1;
                }

                /* Get and print solution of QP. */
                example.getPrimalSolution( Opt );
                q0 = q0 + Opt[0]/10;
                q1 = q1 + Opt[1]/10;
                q_eef = l1 +l2 +l3 + q0 + q1;
                error = p*(q_des-q_eef);
                limit = std::abs(error);

                condition = Opt[0]*A[0] + Opt[1]*A[1] + Opt[2]*A[2];

                printf( "\nOpt = [ %e, %e, %e ]; \n posit= %f, error= %f, q0= %f q1= %f \n",
                        Opt[0],Opt[1],Opt[2],q_eef,error,q0,q1);

                vel_pos.position[0] = q_eef;
                vel_pos.velocity[0] = Opt[0];
                vel_pos.effort[0] = error;
                vel_pos.effort[1] = Opt[1];
                vel_pos.header.stamp  += ros::Duration(0.1);

                joint_pub.publish(vel_pos);

            }
            first = false;
            ros::spinOnce();

        }
        while ( limit>precision &&  i<80 );

        return 0;
    }
}

/* PLOT: rqt_plot /joint_state/position[0] /joint_state/velocity[0]
NOTES:
- Matrix A constant
   - w1 LOW w2 LOW w3 LOW:
   - w1 LOW w2 LOW w3 1:
   - w1 LOW w2 LOW w3 10:
   - w1 LOW w2 1   w3 LOW:
   - w1 LOW w2 1   w3 1:
   - w1 LOW w2 1   w3 10:
   - w1 1   w2 LOW w3 LOW:
   - w1 1   w2 LOW w3 1:
   - w1 1   w2 LOW w3 10:
   - w1 1   w2 1   w3 LOW:
   - w1 1   w2 1   w3 1:
   - w1 1   w2 1   w3 10:



*/

File: examples/dof2.py
#!/usr/bin/env python
# Copyright (c) 2017 Universitaet Bremen - Institute for Artificial Intelligence (Prof. Beetz)
#
# Author: Minerva Gabriela Vargas Gleason <minervavargasg@gmail.com>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY   without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program. If not, see <http://www.gnu.org/licenses/>.

# This program simulates a 2 DOF controller with multiple goals

import rospy
import numpy as np
from qpoases import PyReturnValue as returnValue
from qpoases import PySQProblem as SQProblem
from qpoases import PyOptions as Options
from qpoases import PyPrintLevel as PrintLevel
import plotly
import plotly.graph_objs as go


def main():
    rospy.init_node('controller_2dof', anonymous=True)
    rospy.Rate(10)
    rospy.sleep(0.2)

    # Setup data of QP.
    # Weights.
    w1 = 1e-3
    w2 = 1e-5
    w3 = 1  # slack (attractor)
    # Joint limits.
    q0_max = 0.05
    q1_max = 0.15
    # Links
    l1 = 0.1
    l2 = 0.2
    l3 = 0.3
    # Initial joint values.
    q0_init = q0 = -0.04
    q1_init = q1 = -0.08
    # Joint target.
    q_des = 0.65
    # Slack limits.
    e_max = 1000
    # Velocity limits.(+-)
    v0_max = 0.05
    v1_max = 0.1
    # Acceleration limits.(+-)
    a0_max = 0.05 * 0.5
    a1_max = 0.1 * 0.5
    # Others
    precision = 1e-3
    p = 10
    q_eef_init = q_eef = l1 + l2 + l3 + q0 + q1
    error = p * (q_des - q_eef)
    vel_init = 0
    nWSR = np.array([100])

    # Acceleration
    a0_const = (v0_max - a0_max) / v0_max
    a1_const = (v1_max - a1_max) / v1_max

    example = SQProblem(3, 5)

    H = np.array([w1, 0.0, 0.0,
                  0.0, w2, 0.0,
                  0.0, 0.0, w3]).reshape((3, 3))

    A = np.array([1.0, 1.0, 1.0,
                  1.0, 0.0, 0.0,
                  0.0, 1.0, 0.0,
                  1.0, 0.0, 0.0,
                  0.0, 1.0, 0.0]).reshape((5, 3))

    g = np.array([0.0, 0.0, 0.0, 0.0])
    lb = np.array([-v0_max, -v1_max, -e_max])
    ub = np.array([v0_max, v1_max, e_max])
    lbA = np.array([error, (-q0_max - q0), (-q1_max - q0), -a0_max, -a1_max])
    ubA = np.array([error, (q0_max - q0), (q1_max - q0), a0_max, a1_max])

    # Setting up QProblem object
    options = Options()
    options.setToReliable()
    options.printLevel = PrintLevel.LOW
    example.setOptions(options)

    print("Init pos = %g,  goal = %g, error = %g, q0 =%g, q1 = %g\n" %
          (q_eef, q_des, error, q0, q1))
    print A

    i = 0
    n = 0
    limit = abs(error)
    ok = False
    Opt = np.zeros(3)

    # Plotting
    t = np.array(i)
    pos_eef = np.array(q_eef)
    pos_0 = np.array(q0)
    pos_1 = np.array(q1)
    vel_0 = np.array(vel_init)
    vel_1 = np.array(vel_init)
    vel_eef = np.array(vel_init)
    p_error = np.array(error)
    goal = np.array(q_des)
    lim1 = np.array(q0_max)
    lim2 = np.array(q1_max)

    return_value = example.init(H, g, A, lb, ub, lbA, ubA, nWSR)

    if return_value != returnValue.SUCCESSFUL_RETURN:
        print "Init of QP-Problem returned without success! ERROR MESSAGE: "
        return -1

    while not rospy.is_shutdown():
        while limit > precision and i < 400:
            # Solve QP.
            i += 1
            nWSR = np.array([100])
            lbA[0] = error
            lbA[1] = -q0_max - q0
            lbA[2] = -q1_max - q1
            lbA[3] = a0_const * Opt[0] - a0_max
            lbA[4] = a1_const * Opt[1] - a1_max
            ubA[0] = error
            ubA[1] = q0_max - q0
            ubA[2] = q1_max - q1
            ubA[3] = a0_const * Opt[0] + a0_max
            ubA[4] = a1_const * Opt[1] + a1_max

            return_value = example.hotstart(H, g, A, lb, ub, lbA, ubA, nWSR)

            if return_value != returnValue.SUCCESSFUL_RETURN:
                print "Hotstart of QP-Problem returned without success! ERROR MESSAGE: "
                return -1

            # Get and  print solution of QP.
            example.getPrimalSolution(Opt)
            q0 += Opt[0] / 100
            q1 += Opt[1] / 100
            q_eef = l1 + l2 + l3 + q0 + q1
            error = p * (q_des - q_eef)
            limit = abs(error)

            # print "\nOpt = [ %g, %g, %g, %g ] \n posit= %g, error= %g, q0= %g q1= %g \n" % (
            #    Opt[0], Opt[1], Opt[2], Opt[3], q_eef, error, q0, q1)

            # Plotting arrays
            pos_eef = np.hstack((pos_eef, q_eef))
            pos_0 = np.hstack((pos_0, q0))
            pos_1 = np.hstack((pos_1, q1))
            vel_0 = np.hstack((vel_0, Opt[0]))
            vel_1 = np.hstack((vel_1, Opt[1]))
            vel_eef = np.hstack((vel_eef, Opt[0]+Opt[1]))
            goal = np.hstack((goal, q_des))
            p_error = np.hstack((p_error, error))
            t = np.hstack((t, i))
            lim1 = np.hstack((lim1, q0_max))
            lim2 = np.hstack((lim2, q1_max))

        # Plot
        t_eef = go.Scatter(
            y=pos_eef, x=t, marker=dict(size=4,),
            mode='lines+markers', name='pos_eef')
        t_p0 = go.Scatter(
            y=pos_0, x=t, marker=dict(size=4,),
            mode='lines+markers', name='pos_q0')
        t_p1 = go.Scatter(
            y=pos_1, x=t, marker=dict(size=4,),
            mode='lines+markers', name='pos_q1')
        t_v0 = go.Scatter(
            y=vel_0, x=t, marker=dict(size=4,),
            mode='lines+markers', name='vel_q0')
        t_v1 = go.Scatter(
            y=vel_1, x=t, marker=dict(size=4,),
            mode='lines+markers', name='vel_q1')
        t_q0 = go.Scatter(
            y=lim1, x=t, marker=dict(size=4,),
            mode='lines', name='limit_0')
        t_q1 = go.Scatter(
            y=lim2, x=t, marker=dict(size=4,),
            mode='lines', name='limit_1')
        t_veef = go.Scatter(
            y=vel_eef, x=t, marker=dict(size=4, ),
            mode='lines+markers', name='vel_eef')
        t_er = go.Scatter(
            y=p_error, x=t, marker=dict(size=4,),
            mode='lines+markers', name='error')
        t_goal = go.Scatter(
            y=goal, x=t, marker=dict(size=4, ),
            mode='lines', name='goal')

        print "\n i = %g \n" % (i)

        data_eef = [t_eef, t_veef, t_goal]
        layout_eef = dict(title="Initial position EEF = %g.  Goal = %g, \n" %
                            (q_eef_init, q_des),
                      xaxis=dict(title='Iterations',autotick=False,dtick=25,gridwidth=2,),
                      yaxis=dict(title='Position / Velocity'), )
        fig = dict(data=data_eef, layout=layout_eef)
        plotly.offline.plot(fig, filename='basic_example_eef.html')

        # data = [t_p0, t_p1, t_v0, t_v1, t_q0, t_q1]
        data = [t_p0, t_p1, t_v0, t_v1]
        layout = dict(title="Initial position q0 =%g, q1 = %g.\n" %
                            (q0_init, q1_init),
                      xaxis=dict(title='Iterations',autotick=False,dtick=25,gridwidth=2,),
                      yaxis=dict(title='Position / Velocity'),
                      )
        fig = dict(data=data, layout=layout)
        plotly.offline.plot(fig, filename='basic_example_joints.html')

        return 0


if __name__ == '__main__':
    try:
        main()
    except rospy.ROSInterruptException:
        pass